enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,405 @@
+
+## Proposal
+
+Adopt [MetalLB](https://metallb.universe.tf/) as an out-of-the-box solution for
+most on-premise SLB use cases.
+
+MetalLB is commonly referenced when people discuss service load balancers for
+bare metal.  The [concepts page](https://metallb.universe.tf/concepts/) gives a
+pretty good overview of how it works.  It manages pools of IP addresses to
+allocate for SLBs.  Once an IP is allocated to a SLB, it’s assigned to a Node
+and the location of that IP address must be announced externally.  It has two
+modes to announce IPs: [layer 2](https://metallb.universe.tf/concepts/layer2/)
+(ARP for IPv4, NDP for IPv6) or
+[BGP](https://metallb.universe.tf/concepts/bgp/).  The layer 2 mode is
+sufficient for smaller scale clusters, while the BGP mode can work at much
+larger scale.
+
+While the BGP option is attractive for scaling reasons, it’s also more
+complicated and will not work in all environments.  It won’t work in an
+environment that won’t allow BGP advertisements from the cluster Nodes.  If a
+cluster uses a Network addon that also makes use of BGP, MetalLB integration
+will be more challenging.  For example see the [MetalLB page about Calico
+support](https://metallb.universe.tf/configuration/calico/).
+
+The layer 2 mode has the advantage of working in more environments.  We could
+also consider a MetalLB enhancement that makes it understand different L2
+domains and manage different IP address pools for each domain where SLBs may
+reside.
+
+### User Stories
+
+#### Match Cloud Load Balancer Functionality Where Possible
+
+For each story below which discusses an environment we aim to support, we aim
+to provide functionality that matches what you would see with a cluster on a
+cloud.  This includes allowing Service load balancers accessible inside and
+outside of the cluster, verifiable using existing e2e tests that make use of
+service load balancers.
+
+#### Story 1 - Easy Use with a Small Cluster
+
+As an administrator of a small cluster that resides entirely on a single layer
+2 domain, I would like to configure one or more ranges of IP addresses from my
+network that the cluster is free to use for Service Load Balancers.  I do not
+want to do any extra configuration in my network infrastructure beyond just
+ensuring that the configured ranges of addresses are not used elsewhere.
+
+MetalLB can do this today.
+
+#### Story 2 - BGP Integration
+
+As an administrator of a larger cluster with Nodes that reside on multiple L2
+network segments, I would like to configure one or more ranges of IP addresses
+from my network that the cluster is free to use for Service Load Balancers.  I
+would like my cluster to peer with my BGP infrastructure to advertise the
+current location of IP addresses allocated to Service Load Balancers.
+
+MetalLB can do this today.
+
+#### Story 3 - Larger Clusters without BGP
+
+As an administrator of a larger cluster with Nodes that reside on multiple L2
```

Comment:
I know that the API service is announced using keepalived + HAProxy. I don't know how manual this solution is, however.

Reply:
And you mentioned that below :)

Reply:
the only manual step is specifying two IP addresses at installation time.

Reply:
Yeah, I think this is discussed in the Alternatives section